browser/brave_news/brave_news_tab_helper.cc

```diff
@@ -0,0 +1,159 @@
+#include "content/public/browser/browser_context.h"
+#include "content/public/browser/browser_thread.h"
+#include "content/public/browser/page.h"
+#include "content/public/browser/web_contents.h"
+#include "content/public/browser/web_contents_observer.h"
+#include "content/public/browser/web_contents_user_data.h"
+
+BraveNewsTabHelper::BraveNewsTabHelper(content::WebContents* contents)
+    : content::WebContentsUserData<BraveNewsTabHelper>(*contents),
+      content::WebContentsObserver(contents),
+      controller_(
+          brave_news::BraveNewsControllerFactory::GetControllerForContext(
+              contents->GetBrowserContext())) {
+  controller_->publisher_controller()->AddObserver(this);
+  controller_->GetPublishers(base::DoNothing());
+}
+
+BraveNewsTabHelper::~BraveNewsTabHelper() {
+  controller_->publisher_controller()->RemoveObserver(this);
+}
+
+std::vector<BraveNewsTabHelper::FeedDetails>
+BraveNewsTabHelper::GetAvailableFeeds() {
+  std::vector<FeedDetails> feeds;
+
+  std::unordered_set<std::string> seen_feeds;
+  auto default_publisher =
+      controller_->publisher_controller()->GetPublisherForSite(
+          GetWebContents().GetLastCommittedURL());
+  if (default_publisher) {
+    seen_feeds.insert(default_publisher->feed_source.spec());
+    feeds.insert(feeds.begin(), {default_publisher->feed_source,
+                                 default_publisher->publisher_name});
+  }
+
+  for (const auto& rss_feed : rss_page_feeds_) {
+    auto url = rss_feed.feed_url.spec();
+    if (seen_feeds.find(url) != seen_feeds.end())
+      continue;
+
+    seen_feeds.insert(url);
+    feeds.push_back(rss_feed);
+  }
+
+  return feeds;
+}
+
+bool BraveNewsTabHelper::is_subscribed(const FeedDetails& feed_details) {
+  auto publisher = controller_->publisher_controller()->GetPublisherForFeed(
+      feed_details.feed_url);
+  return brave_news::IsPublisherEnabled(publisher);
+}
+
+bool BraveNewsTabHelper::is_subscribed() {
+  for (const auto& feed : GetAvailableFeeds()) {
+    if (is_subscribed(feed))
+      return true;
+  }
+  return false;
+}
+
+void BraveNewsTabHelper::ToggleSubscription(const FeedDetails& feed_details) {
+  bool subscribed = is_subscribed(feed_details);
+  auto publisher = controller_->publisher_controller()->GetPublisherForFeed(
+      feed_details.feed_url);
+  if (publisher) {
+    controller_->SetPublisherPref(
+        publisher->publisher_id, subscribed
+                                     ? brave_news::mojom::UserEnabled::DISABLED
+                                     : brave_news::mojom::UserEnabled::ENABLED);
+  } else if (!subscribed) {
+    controller_->SubscribeToNewDirectFeed(feed_details.feed_url,
+                                          base::DoNothing());
+  }
+}
+
+void BraveNewsTabHelper::OnReceivedRssUrls(const GURL& site_url,
+                                           std::vector<GURL> feed_urls) {
+  if (site_url != GetWebContents().GetLastCommittedURL())
```

Comment:
I think it can, as the WebContents could navigate before we receive the RSS urls (i.e, we navigate but have only just received the URLs), though maybe the navigation would cancel the RSS urls request?

Reply:
Oops, I think I wanted to comment above to `BraveNewsTabHelper::OnFoundFeeds()` not for `OnReceivedRssUrls()`.

Reply:
Ah, got it NVM. it's same :)